zipkin-server/src/main/java/zipkin2/server/internal/throttle/ThrottledCall.java

```diff
@@ -37,183 +37,187 @@
+   * is ok because in almost all cases, doing so would imply invoking {@link #enqueue(Callback)}
+   * anyway.
+   */
+  @Override protected Void doExecute() throws IOException {
+    AwaitableCallback awaitableCallback = new AwaitableCallback();
+    doEnqueue(awaitableCallback);
+    if (!await(awaitableCallback.countDown)) throw new InterruptedIOException();
+
+    Throwable t = awaitableCallback.throwable;
+    if (t != null) {
+      if (t instanceof Error) throw (Error) t;
+      if (t instanceof IOException) throw (IOException) t;
+      if (t instanceof RuntimeException) throw (RuntimeException) t;
+      throw new RuntimeException(t);
     }
+    return null; // Void
   }
 
-  @Override protected void doEnqueue(Callback<V> callback) {
-    Listener limitListener = limiter.acquire(null)
+  @Override protected void doEnqueue(Callback<Void> callback) {
+    Listener limiterListener = limiter.acquire(null)
       .orElseThrow(RejectedExecutionException::new); // TODO: make an exception message
+    limiterMetrics.requests.increment();
+    limiterListener = limiterMetrics.wrap(limiterListener);
+
+    LimiterReleasingCallback releasingCallback =
+      new LimiterReleasingCallback(callback, isOverCapacity, limiterListener);
 
     try {
-      executor.execute(new QueuedCall<>(this, callback, limitListener));
-    } catch (RuntimeException | Error e) {
-      propagateIfFatal(e);
-      // Ignoring in all cases here because storage itself isn't saying we need to throttle.  Though, we may still be
-      // write bound, but a drop in concurrency won't necessarily help.
-      limitListener.onIgnore();
-      throw e;
+      executor.execute(new EnqueueAndAwait(this, releasingCallback));
+    } catch (RuntimeException | Error t) { // possibly rejected, but from the executor, not storage!
+      propagateIfFatal(t);
+      callback.onError(t);
+      // Ignoring in all cases here because storage itself isn't saying we need to throttle. Though
+      // we may still be write bound, but a drop in concurrency won't necessarily help.
+      limiterListener.onIgnore();
+      throw t;
     }
   }
 
-  @Override public Call<V> clone() {
-    return new ThrottledCall<>(executor, limiter, isOverCapacity, delegate.clone());
+  @Override public Call<Void> clone() {
+    return new ThrottledCall(delegate.clone(), executor, limiter, limiterMetrics, isOverCapacity);
   }
 
   @Override public String toString() {
     return "Throttled(" + delegate + ")";
   }
 
-  static String setCurrentThreadName(String name) {
-    Thread thread = Thread.currentThread();
-    String originalName = thread.getName();
-    thread.setName(name);
-    return originalName;
+  static final class AwaitableCallback implements Callback<Void> {
+    final CountDownLatch countDown = new CountDownLatch(1);
+    Throwable throwable; // thread visibility guaranteed by the countdown latch
```

Comment:
Good comment, it'd be unclear without I think.

Incidentally, realized it didn't matter here so much - a `volatile` variable has equivalent performance to a synchronized one if it's only ever written / read once ;)